### PR TITLE
refactor: self distillation trainers (sdpo/sdft/...) 

### PR DESCRIPTION
**Pull request by LeonEricsson**
_fyi see #5609 for a concrete extension (SD-Zero) on top of this PRs refactor_

### Summary

Major internal restructuring of the experimental self-distillation trainers. The core change: unify shared lifecycle logic in `BaseSelfDistillationTrainer`, while concrete trainers (`SDFTTrainer`, `SDPOTrainer`) own their algorithmic recipes.

The previous design mixed generic self-distillation lifecycle with SDPO-specific reward/scoring machinery in the same inheritance stack, meaning the base implicitly assumed every method needed reward-based optimization. This PR corrects that with a clean split: base owns infrastructure, trainers own semantics. 

No intended change in functionality.

### What changed

**Base class now owns the common lifecycle:**
model/tokenizer/generation setup, student-side rollout generation (transformers + vLLM), and teacher synchronization.

**Concrete trainers own the method recipe:**
rollout interpretation, teacher input construction, algorithm-specific metadata (rewards, advantages, etc.), and loss composition.

**SDPO owns its identity.**
Reward scoring and policy optimization are no longer baked into the base. SDPO explicitly manages its own reward functions, normalization, and hybrid loss.

**Teacher regularization**
is now a single semantic config parameter (`teacher_model_kind`: `"live"` / `"base"` / `"ema"`), with the base handling PEFT/non-PEFT details internally.

### Removed

I’d like to touch on a few abstractions I’ve cleared out and the reasoning behind those calls. In short, I felt that **`PromptTokenizer`**, **`SelfDistillationMixin`**, and **`OnlineRolloutMixin`** were "leaky"- masking the logic they were supposed to define and making the architecture harder to follow. 

`PromptTokenizer` logic now lives directly in the base, next to the rollout pipeline it serves. The two mixins were splitting orchestration, rollout mechanics, and algorithm-specific tasks (like reward scoring) across too many pieces. That responsibility is now split cleanly: **`BaseSelfDistillationTrainer`** handles the "heavy lifting" of the distillation machinery and rollout mechanics. Meanwhile, concrete trainers have full ownership over their specific batch finalization and loss logic.

### Caveat

- `generate_from_teacher` isn't ported; the base currently assumes student-sourced rollouts. There's still hanging code that references it, so we need to decide whether this is a capability worth porting before this PR can land.

- As per discussion on slack, this PR still retains the Base class contract, forcing trainer (SDPO,SDFT) coupling. While this works well for SDPO, SDFT and SD Zero, long term stability of these contracts is uncertain. An alternate path is to drop the Base altogether, let each trainer live separately, and move common functionality into pure helpers.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

---

> [!NOTE]
> **High Risk**
> Large refactor of core experimental training infrastructure (rollout generation, teacher syncing, loss computation, and SDPO reward/policy logic) with potential behavior/regression risk across training/eval flows, especially around PEFT teacher selection and new distillation modes.
> 
> **Overview**
> **Refactors the experimental self-distillation trainers** by moving shared lifecycle responsibilities (prompt tokenization, rollout generation for transformers/vLLM, teacher model selection/sync, distillation logits/loss plumbing, metrics/logging) into `BaseSelfDistillationTrainer`, and deleting the old `OnlineRolloutMixin`/`SelfDistillationMixin` layering.
> 
> **Updates SDPO and SDFT to own their method-specific “recipes”.** `SDPOTrainer` now explicitly manages reward model/function setup, reward normalization/advantages, and policy vs distillation loss composition (adds `policy_only` mode), while `SDFTTrainer` is ported onto the new base and builds teacher prompts/masks during `finalize_batch` (with `generate_from_teacher` explicitly unsupported).
> 
> **Modernizes configuration and docs/tests**: introduces explicit `distillation_mode` (`sampled_token`/`full_logits`/`topk_logits`) replacing `full_logit_distillation`, adds semantic `teacher_model_kind` (`base`/`live`/`ema`) with `teacher_sync_steps`, and updates docs/examples/tests accordingly; adds new unit coverage for base trainer tokenization, teacher-kind behavior, and distillation loss masking/IS clipping.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit bca1a77fa5084e8eb236f5c056c67533e2e42ab3. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/self_distillation/base_self_distillation_trainer.py [hunk 1]
```diff
@@ -212,75 +260,113 @@ def __init__(
...
+                callback_fn(
+                    args=self.args,
+                    state=self.state,
+                    control=self.control,
+                    model=self.model,
+                    processing_class=self.processing_class,
+                    **payload,
+                )
+
+    def _setup_teacher_model(self) -> None:
```
Comment (LeonEricsson) [inline trl/experimental/self_distillation/base_self_distillation_trainer.py:292]: This PR promotes LoRA and teacher regularization into shared capabilities owned by the base, rather than being method-specific. Teacher handling is collapsed into the semantic `teacher_model_kind` API (`live` / `base` / `ema`), letting the base infer PEFT implementation details.

File: trl/experimental/self_distillation/base_self_distillation_trainer.py [hunk 2]
```diff
@@ -333,24 +419,435 @@ def training_step(self, model, inputs, num_items_in_batch):
...
                 self._dispatch_self_distillation_callback(
                     "on_generation_batch_built",
                     generate_every=generate_every,
                     steps_per_generation=self.args.steps_per_generation,
                 )
             return self._buffered_inputs[self._step % self.args.steps_per_generation]
-        return self._build_buffered_batch(generation_batch)
+        return self._prepare_training_batch(generation_batch)
+
+    def _prepare_training_batch(self, inputs: list[dict[str, Any]]) -> TrainingBatch:
```
Comment (LeonEricsson) [inline trl/experimental/self_distillation/base_self_distillation_trainer.py:451]: This is the hinge of the refactor. The base owns `sample_rollouts()` and only asks subclasses to implement `finalize_batch(...)`, which takes a `RolloutBatch` (prompts + completions, pre-teacher) and produces a `TrainingBatch` with teacher-conditioned inputs and any trainer-specific metadata. Everything downstream keys off this split

File: trl/experimental/self_distillation/base_self_distillation_trainer.py [hunk 3]
```diff
@@ -333,24 +419,435 @@ def training_step(self, model, inputs, num_items_in_batch):
...
+                tokenize=True,
+                return_dict=True,
+                **self.chat_template_kwargs,
+            )
+            prompt_ids = tokenized["input_ids"]
+        else:
+            prompt_ids = self.processing_class(text=prompts)["input_ids"]
+        return prompt_ids
+
+    def _tokenize_prompts(self, prompts: list[Any]) -> list[list[int]]:
```
Comment (LeonEricsson) [inline trl/experimental/self_distillation/base_self_distillation_trainer.py:482]: Two things regarding prompt tokenization:
1. **Removed `PromptTokenizer`.** Prompt rendering used to live in a separate `PromptTokenizer` object the trainer held and builders reached back through. It had no stable contract and was effectively just forwarding trainer state. Tokenization now lives on the base.
2. **Correctness fix on the rollout path.** Prompts are now tokenized directly into token ID sequences, and generation runs from those for both transformers and vLLM, rather than round-tripping through text. This matches what existing trainers already do and eliminates prompt → text → token mismatches.

File: trl/experimental/sdft/sdft_trainer.py
Comment (LeonEricsson) [inline trl/experimental/sdft/sdft_trainer.py:1]: Under the new base, SDFT's trainer becomes very simple. Build the teacher-conditioned batch and define loss composition.

Comment (LeonEricsson) [inline trl/experimental/sdft/sdft_trainer.py:93]: _generate_from_teacher stuff that needs to be addressed

File: trl/experimental/sdpo/sdpo_trainer.py [hunk 4]
```diff
@@ -277,44 +263,194 @@ def __init__(
...
+                rewards_per_func[:, i] = torch.tensor(output_reward_func, dtype=torch.float32, device=device)
 
-    def _allow_topk_without_full_logit_distillation(self) -> bool:
-        return False
+        return self.accelerator.gather(rewards_per_func)
 
-    def _generate_and_score_completions(
-        self, inputs: list[dict[str, torch.Tensor | Any]]
-    ) -> dict[str, torch.Tensor | Any]:
+    def finalize_batch(
```
Comment (LeonEricsson) [inline trl/experimental/sdpo/sdpo_trainer.py:380]: SDPO's  counterpart to the new base boundary. The base provides the generic rollout batch, and SDPO’s `finalize_batch(...)` owns everything that is actually SDPO-specific.

File: trl/experimental/self_distillation/base_self_distillation_trainer.py [hunk 5]
```diff
@@ -67,18 +74,59 @@
...
+    """Aligned logits and masks used to compute a self-distillation objective."""
+
+    completion_ids: torch.Tensor
+    completion_mask: torch.Tensor
+    response_mask: torch.Tensor
+    student_logits: torch.Tensor
+    teacher_logits: torch.Tensor
+
+
+class BaseSelfDistillationTrainer(_BaseTrainer, ABC):
```
Comment (LeonEricsson) [inline trl/experimental/self_distillation/base_self_distillation_trainer.py:117]: The new base class deliberately owns *capability only*: model/tokenizer setup, rollout generation, teacher sync infrastructure, shared distillation math. It does **not** own any algorithmic choices, no reward scoring, no advantage computation, no loss composition. Those are pushed into subclasses.

File: trl/experimental/self_distillation/base_self_distillation_trainer.py [hunk 6]
```diff
@@ -333,24 +420,423 @@ def training_step(self, model, inputs, num_items_in_batch):
...
+                    self.model,
+                    prompt_completion_ids,
+                    attention_mask,
+                    logits_to_keep,
+                )
+                old_per_token_logps = select_token_log_probs(logits, completion_ids)
+
+        return old_per_token_logps
+
+    def _compute_self_distillation_loss(
```
Comment (LeonEricsson) [inline trl/experimental/self_distillation/base_self_distillation_trainer.py:612]: This loss computation is opinionated: it covers top-k / full-logit / sampled-token paths, importance-sampling clipping, and aggregation per `loss_type`, matching what both SDPO and SDFT need. It lives on the base because both current trainers use it as-is, and any future trainer doing standard self-distillation math gets it for free.

That said, it's not part of the base contract. `_compute_self_distillation_loss` is only ever called from a trainer's own `compute_loss`, and nothing in the base touches it. A trainer with substantially different distillation math can just ignore it and compute loss from scratch.

**Comment by LeonEricsson** [comment]
I've provided a bunch of comments to aid reviewers
